File: Flyweight/Flyweight.php
<?php

/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/23
 * Time: 23:22
 */
interface Shape
{
    public function draw();
}

class Circle implements Shape
{
    private $color, $x, $y, $radius;

    public function __construct($color)
    {
        $this->color = $color;
    }

    public function setX($x)
    {
        $this->x = $x;
    }


    public function setY($y)
    {
        $this->y = $y;
    }

    public function setRadius($radius)
    {
        $this->radius = $radius;
    }

    public function draw()
    {
        // TODO: Implement draw() method.
        var_dump("Circle: Draw() [Color : ".$this->color.", x : ".$this->x.", y :".$this->y.", radius :".$this->radius);
    }
}

class ShapeFactory
{
    private static $circleMap = array();

    public static function getCircle($color)
    {

        if(isset(self::$circleMap[$color])){
            $circle = self::$circleMap[$color];
        }else{
            $circle = new Circle($color);
            self::$circleMap[$color] = $circle;
            var_dump("Creating circle of color : ".$color);
        }
        return $circle;
    }

}


//demo
$colors=array("Red", "Green", "Blue", "White", "Black");
for($i=0; $i < 20; ++$i) {
    $randColor=$colors[array_rand($colors)];
    $circle=ShapeFactory::getCircle($randColor);
    $circle->setX(rand(50,100));
    $circle->setY(rand(50,100));
    $circle->setRadius(100);
    $circle->draw();
}

File: Flyweight/Flyweight.go
package main

import (
	"fmt"
	"math/rand"
)

type Shape interface {
	draw()
}

type Circle struct {
	color  string
	x      int
	y      int
	radius int
}

func (c *Circle)setX(x int) {
	c.x = x
}

func (c *Circle)setY(y int) {
	c.y = y
}

func (c *Circle)setRadius(radius int) {
	c.radius = radius
}

func (c *Circle)draw() {
	fmt.Println("Circle: Draw() [Color : ", c.color, ", x : ", c.x, ", y :", c.y, ", radius :", c.radius)
}

type ShapeFactory struct {
	shapeMap map[string]*Circle
}

func(s *ShapeFactory)init(){
	s.shapeMap=make(map[string]*Circle)
}

func (s *ShapeFactory)getCircle(color string) *Circle {
	shape := s.shapeMap[color];
	if shape == nil{
		shape = &Circle{color:color}
		s.shapeMap[color]=shape
		fmt.Println("Creating circle of color : ",color)
	}
	return shape
}

func main() {
	colors := []string{"Red", "Green", "Blue", "White", "Black"}
	shapeFactory := new(ShapeFactory)
	shapeFactory.init()
	for i := 1; i <= 20; i++ {
		cn := rand.Intn(5);
		circle := shapeFactory.getCircle(colors[cn])
		circle.setX(rand.Intn(100))
		circle.setY(rand.Intn(100))
		circle.setRadius(100)
		circle.draw()

	}

}

